Identifer validate ISBN

BioFreak95 opened the pull request:
Hi guys,

I add an validator for identifier, who delegates the validation to the validator-class of the chosen identifier.

j3nsch commented on library/Application/Form/Validate/IdentifierValidate.php:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Class for delegate validator for identifier in Admin-Form.
+ */
+class Application_Form_Validate_IdentifierValidate extends Zend_Validate_Abstract
```
Please, change the name of the class to "Application_Form_Validate_Identifier". Having "Validate" twice in the name seems unnecessary.

j3nsch commented on modules/admin/forms/Document/Identifier.php:
```diff
@@ -59,63 +60,69 @@ class Admin_Form_Document_Identifier extends Admin_Form_AbstractModelSubForm {
-        $this->addElement('text', self::ELEMENT_VALUE, array('required' => true, 'size' => '80'));
+        $typeElement = $this->createElement('Identifier', self::ELEMENT_TYPE, array('required' => true));
+        $this->addElement($typeElement);
+        $valueElement = $this->createElement('text', self::ELEMENT_VALUE, array('required' => true, 'size' => '80'));
```
Please, limit the length of lines to 120 characters.

j3nsch commented on tests/library/Application/Form/Validate/IdentifierValidateTest.php:
```diff
@@ -0,0 +1,74 @@
+        $this->assertFalse($this->_validator->isValid('978-3-86680-192-13'));
+    }
+
+    public function testIsValidEmpty() {
```
The name of the function does not describe the test properly. It is about having a type other than "ISBN".

BioFreak95 commented:
It should be much better now.
I changed the names for the classes
I wrote more tests (10 now)
I wrote comments now for my code
And I changed some style, so the code should be better formatted

j3nsch commented:
Ich bin gerade bei Ticket OPUSVIER-3479 in dem es um "ehrlichere" Tests geht. In Zukunft sollen alle Unit Tests mit "@covers" sagen, welche Funktion sie testen. In der PHPUnit Dokumentation ist das genauer erklärt. Für die Klasse kann "@coversDefaultClass Application_Form_Validate_Identifier" verwendet werden und bei den Testfunktionen dann "@covers isValid". Damit wird nur Coverage für die `isValid` Funktion erzeugt. Dazu müssen wir noch etwas in der Entwicklerdokumentation schreiben.

j3nsch commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+
+    }
+
+
```
Ein `return` am Ende der Funktion ist gute Praxis. In diesem Fall könnte es zum Beispiel **false**  sein. Dann könnte der "default"-Zweig oben entsprechend verändert werden, so dass er gegebenenfalls **true** zurück gibt und ansonsten die Funktion bis zum Ende ausgeführt wird.

BioFreak95 replied:
Have an idea, but we have to discuss before implementation

BioFreak95 replied:
check

j3nsch commented on modules/admin/forms/Document/Identifier.php:
```diff
@@ -59,63 +60,70 @@ class Admin_Form_Document_Identifier extends Admin_Form_AbstractModelSubForm {
+        $typeElement = $this->createElement('Identifier', self::ELEMENT_TYPE, array('required' => true));
+        $this->addElement($typeElement);
+        $valueElement = $this->createElement('text', self::ELEMENT_VALUE, array(
+            'required' => true, 'size' => '80'));
```
Bitte "));" am Ende in die nächste Zeile.

BioFreak95 replied:
We need to talk about this comment

BioFreak95 replied:
check

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+
+/**
+ * Unittests for  Class Application_Form_Validate_IdentifierTest
+ */
```
Bitte "@covers Application_Form_Validate_Identifier" einfügen, um Coverage zu beschränken.

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    /**
+     * Test for an false argument in an ISBN-identifier
```
Kurzbeschreibung (erster Satz im Kommentar) bitte immer mit Punkt "." abschließen.

BioFreak95 replied:
check

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+        $this->assertFalse($this->_validator->isValid(''));
+    }
+    /**
+     * Test for an true argument in an ISBN-identifier
```
Kommentar passt nicht ganz.

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Test for an NULL argument in an ISBN-identifier
+     */
+    public function testIsValidISBNNULL() {
```
Bitte vorsicht mit Großbuchstaben. Hier wäre "testIsValidIsbnNull" besser auch wenn ISBN sonst nur mit Großbuchstaben geschrieben wird.

BioFreak95 replied:
check

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Test for an true argument in an ISBN-identifier
+     */
+    public function testIsValidTrueArgument() {
```
"testIsValidTrue" würde reichen.

BioFreak95 replied:
check

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+        $this->_validator = new Application_Form_Validate_Identifier($this->_element);
+        $this->assertFalse($this->_validator->isValid(''));
+    }
+}
```
Versuch mal bitte noch zu testen, ob die Nachrichten korrekt sind, ob sie also von der anderen Validierungsklasse übernommen werden und ob sie zum Validierungsfehler passen.

BioFreak95 replied:
Check

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+     * Test for an wrong ISBN in an ISBN-identifier
+     */
+    public function testIsValidWrongISBN() {
+        $this->assertFalse($this->_validator->isValid('978-3-86680-192-13'));
```
Es gibt unterschiedliche Gründe warum die ISBN nicht gültig ist, z.B. Formatierung und Prüfsumme. Möglichst alle Fälle abdecken. Am Besten auch immer mit mindestens zwei Werten testen.

BioFreak95 replied:
check

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Unittests for  Class Application_Form_Validate_IdentifierTest
+ */
+class Application_Form_Validate_IdentifierTest extends ControllerTestCase
```
Für den Testcode gelten die selben Regeln beim Coding Style. Also zum Beispiel Leerzeilen zwischen den Funktionen.

BioFreak95 replied:
check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * @param mixed $value Element
+     * @return bool
+     * Check which type the identifier-element has and delegate at the validator of the type or return true, if the validator
```
We need to talk about this comment

BioFreak95 replied:
check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+ */
+class Application_Form_Validate_Identifier extends Zend_Validate_Abstract
+{
+    /**
```
Check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+
+    /**
+     * Application_Form_Validate_Identifier constructor.
+     * @param Opus_Identifier $element
```
Check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * @param mixed $value Element
+     * @return bool
+     * Check which type the identifier-element has and delegate at the validator of the type or return true, if the validator
```
Check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+        $value = (string)$value;
+        $this->_setValue($value);
+
+        switch (strtoupper($this->_element->getValue())) {
```
Check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+                }
+
+        }
+
```
Check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+
+
+    }
+
```
Check

j3nsch commented on modules/admin/forms/Document/Identifier.php:
```diff
@@ -59,63 +60,70 @@ class Admin_Form_Document_Identifier extends Admin_Form_AbstractModelSubForm {
+        $this->addElement($typeElement);
+        $valueElement = $this->createElement('text', self::ELEMENT_VALUE, array(
+            'required' => true, 'size' => '80'));
+        $valueElement->addValidator(new Application_Form_Validate_IdentifierValidate($typeElement));
```
BUG! Die Klasse `Application_Form_Validate_Identifier` wurde umbenannt. Hier steht noch der alte Name, was zu einem Fehler bei der Ausführung des Formulars führt.

BioFreak95 replied:
Check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+            case 'ISBN':
+                $validateISBN = new Opus_Validate_Isbn();
+                $result = $validateISBN->isValid($value);
+                $this->_messageTemplates = $validateISBN->getMessageTemplates();
```
check

BioFreak95 commented on library/Application/Form/Validate/Identifier.php:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * @param mixed $value Element
```
check

BioFreak95 commented:
If there are some more points, you like to have a change, please let me know.

j3nsch commented on tests/library/Application/Form/Validate/IdentifierTest.php:
```diff
@@ -0,0 +1,211 @@
+    {
+        $this->_validator = new Application_Form_Validate_Identifier(null);
+    }
+
```
Hier fehlt noch ein Test, wenn etwas übergeben wird was nicht vom Typ Zend_Form_Element ist.

BioFreak95 replied:
Habe noch zwei weitere Tests zu der Übergabe geschrieben.
Wenn du noch was hast, sag einfach Bescheid. Lieber am Anfang Perfektionismus als später dann Probleme zu haben.
